pallets/clusters/src/lib.rs

```diff
@@ -0,0 +1,520 @@
+					let role = cluster
+						.roles
+						.get(index)
+						.expect("Should find the cluster");
+					let mut members = role.clone().members;
+					for member in new_members_list {
+						members.push(member);
+					}
+				});
+			});
+
+			<RoleToSettings<T>>::mutate(&role_hash, |role| {
+				let role_settings = role
+					.as_mut()
+					.expect("Should find the role settings");
+				role_settings.push(new_settings);
+			});
+
+			Self::deposit_event(Event::RoleEdited { role_hash });
+			log::trace!("Role edited: {:?}", role_hash);
+
+			Ok(())
+		}
+
+		/// Delete a **Role**.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn delete_role(
+			origin: OriginFor<T>,
+			role_name: BoundedVec<u8, T::NameLimit>,
+			cluster_id: Hash128,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!role_name.len().is_zero(), Error::<T>::InvalidInput);
+
+			// only the owner of the cluster can do this operation
+			let cluster = <Clusters<T>>::get(&cluster_id).ok_or(Error::<T>::ClusterNotFound)?;
+			ensure!(who == cluster.owner, Error::<T>::NoPermission);
+
+			let role_name_vec = role_name.into_inner();
+
+			// Check that the role exists in the cluster and in storage
+			if !cluster
+				.roles
+				.iter()
+				.any(|role| role_name_vec.eq(&role.name))
+			{
+				return Err(Error::<T>::RoleNotFound.into())
+			}
+
+			// write
+			// Remove Role from Cluster
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster
+					.as_mut()
+					.expect("Should find the cluster");
+				cluster
+					.roles
+					.iter()
+					.position(|x| x.name == role_name_vec)
+					.map(|index| cluster.roles.remove(index));
+			});
+
+			let role_hash = blake2_128(&[&cluster_id[..], &role_name_vec.as_slice()].concat());
+			<RoleToSettings<T>>::remove(&role_hash);
+
+			Self::deposit_event(Event::RoleDeleted { role_hash });
+			log::trace!("Role deleted: {:?}", role_hash);
+
+			Ok(())
+		}
+
+		/// Create a **Member**, give it a set of existing **Role** and assign it to an existing **Cluster**.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn add_member(
+			origin: OriginFor<T>,
+			cluster_id: Hash128,
+			roles: Vec<Vec<u8>>,
+			member: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// Check that the cluster exists and the caller is the owner of the cluster
+			let cluster = <Clusters<T>>::get(cluster_id).ok_or(Error::<T>::ClusterNotFound)?;
+			ensure!(who == cluster.owner, Error::<T>::NoPermission);
+
+			// Check that the cluster does not already contain the member
+			ensure!(!cluster.members.contains(&member), Error::<T>::MemberExists);
+
+			ensure!(
+				cluster.members.len() < T::MembersLimit::get() as usize,
+				Error::<T>::MembersLimitReached
+			);
+
+			// Check that the roles for the member already exists in the cluster
+			let roles_in_cluster: Vec<Vec<u8>> =
```

Comment:
why? to just ensure `ensure!(roles_in_cluster.contains(&role), Error::<T>::RoleNotFound);` you don't need this

Reply:
I'm wrong, it's multiple roles